testplan/common/entity/base.py

```diff
@@ -96,10 +101,14 @@ def __getattr__(self, item):
             if item in self.parent.cfg.initial_context:
                 return self.parent.cfg.initial_context[item]
 
-        return self.__getattribute__(item)
+        raise AttributeError(
+            "'{}' object has no attribute '{}'".format(
+                self.__class__.__name__, item
+            )
+        )
 
     def __getitem__(self, item):
-        return getattr(self, item)
+        return self.__getattr__(item)
```

Comment:
Mark: in context resolving this work, but still `env.abc` cannot work in testcases.

Reply:
what if the add would be enhanced to do sanity checks and refuse to add drivers whose name clash with members/inital context keys?

Reply:
`Environment` class has many members (although only `logger`, `first`, `parent` might be used as driver name). So I think the agent class `RuntimeEnvironment` for user's testcase can help to fetch the driver correctly, while Testplan internal code mostly uses `Environment`. There could still be name conflicts (fortunately `RuntimeEnvironment` only has 2 menbers `_environment` and `runtime_info`)

Reply:
> what if the add would be enhanced to do sanity checks and refuse to add drivers whose name clash with members/inital context keys?

I would like to have that too!

Reply:
I think that in `add` we need to check if the `uid` already in resources and parent's initial_context (we have not done this and I'll add some code for it), but it's not smart to check duplicates of its own member, unless we can make sure from now to the future this class is fixed and only have several members, just like we check testcase's name in `suite.py`.